### PR TITLE
RavenDB-18116 abstract subscription connection & state

### PR DESCRIPTION
**Pull request by karmeli87**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-18116

### Additional description

- abstract subscription connection & state so we can reuse the code for the orchestrated connection & state
- use the the Topology of the database record for the sharded whosTaskIsIt subscription
- centralize the sharded subscription to reside on the database context

### CODE REVIEW

File: src/Raven.Server/Documents/Handlers/Processors/Subscriptions/AbstractSubscriptionsHandlerProcessorForGetSubscription.cs [hunk 1]
```diff
@@ -114,23 +114,23 @@ private static DynamicJsonValue GetSubscriptionConnectionJson(SubscriptionConnec
             {
                 [nameof(SubscriptionConnection.ClientUri)] = x.ClientUri,
                 [nameof(SubscriptionConnection.Strategy)] = x.Strategy,
-                [nameof(SubscriptionConnection.Stats)] = GetConnectionStatsJson(x.Stats),
+                [nameof(SubscriptionConnection.StatsCollector)] = GetConnectionStatsJson(x.StatsCollector),
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Handlers/Processors/Subscriptions/AbstractSubscriptionsHandlerProcessorForGetSubscription.cs]: do we need change this?

File: src/Raven.Server/Documents/Sharding/ShardedDatabaseContext.Subscriptions.cs [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+{
+    public ShardedSubscriptions Subscriptions;
+
+    public class ShardedSubscriptions : ISubscriptionSemaphore
+    {
+        private readonly ShardedDatabaseContext _context;
+        private readonly ServerStore _serverStore;
+        private readonly string _databaseName;
+
+        public readonly SemaphoreSlim ConcurrentConnectionsSemiSemaphore;
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Sharding/ShardedDatabaseContext.Subscriptions.cs:25]: semisemaphore?

File: src/Raven.Server/Documents/Sharding/Subscriptions/ShardedSubscriptionWorker.cs [hunk 3]
```diff
@@ -123,5 +109,24 @@ internal override async Task ProcessSubscriptionInternalAsync(JsonContextPool co
                 }
             }
         }
+
+        private async Task TryPublishBatch(BatchFromServer incomingBatch)
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Sharding/Subscriptions/ShardedSubscriptionWorker.cs]: TryPublishBatchAsync

File: src/Raven.Server/Documents/Subscriptions/SubscriptionBinder.cs [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+        if (tcpConnectionOptions.DocumentDatabase is ShardedDocumentDatabase shardedDocumentDatabase)
+        {
+            var connectionForShard = new SubscriptionConnectionForShard(server, tcpConnectionOptions, onDispose, buffer,
+                shardedDocumentDatabase.ShardedDatabaseName);
+            connection = connectionForShard;
+            return new SubscriptionBinder<SubscriptionConnectionsState, SubscriptionConnection>(tcpConnectionOptions.DocumentDatabase.SubscriptionStorage,
+                new Lazy<SubscriptionConnectionsState>(connectionForShard.GetSubscriptionConnectionStateForShard), connectionForShard);
+        }
+
+        var connection2 = new SubscriptionConnection(server, tcpConnectionOptions, onDispose, buffer,
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Subscriptions/SubscriptionBinder.cs]: connection2?

File: src/Raven.Server/Documents/Subscriptions/SubscriptionConnectionsState.cs [hunk 5]
```diff
@@ -1,514 +1,70 @@
...
-using Sparrow.Collections;
 using Sparrow.Server;
-using Sparrow.Server.Utils;
-using Sparrow.Threading;
-using Voron;
 
 namespace Raven.Server.Documents.Subscriptions
 {
-    public class SubscriptionConnectionsState : IDisposable
+    public class DummySubscriptionConnectionsState : SubscriptionConnectionsState
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Subscriptions/SubscriptionConnectionsState.cs]: move o another file

File: src/Raven.Server/Documents/Subscriptions/SubscriptionConnectionsState.cs [hunk 6]
```diff
@@ -1,514 +1,70 @@
...
-                            $"Subscription {currentConnection?.Options.SubscriptionName} does not accept concurrent connections at the moment, connection cannot be opened");
-                    default:
-                        throw new InvalidOperationException("Unknown subscription open strategy: " +
-                                                            incomingConnection.Strategy);
-                }
-            }
-
-            RegisterOneConnection(incomingConnection);
+            var (etag, _) = await _server.SendToLeaderAsync(command);
+            await WaitForIndexNotification(etag);
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Subscriptions/SubscriptionConnectionsState.cs]: Async

File: test/StressTests/Rachis/SubscriptionsFailoverStress.cs [hunk 7]
```diff
@@ -74,20 +74,20 @@ private static async Task ModifyDocuments(IDocumentStore store, int start)
             }
         }
 
-        [Fact]
-        public async Task SubscriptionFailoverWhileModifying()
+        [RavenTheory(RavenTestCategory.Subscriptions)]
+        [RavenData(DatabaseMode = RavenDatabaseMode.Single)]
+        [RavenData(DatabaseMode = RavenDatabaseMode.Sharded, Skip = "Fix me")]
```
Comment (ppekrol) [inline test/StressTests/Rachis/SubscriptionsFailoverStress.cs]: was issue created in YT?

File: src/Raven.Server/Documents/Sharding/ShardedDatabaseContext.Subscriptions.cs (hunk 2 above)
Comment (karmeli87) [reply src/Raven.Server/Documents/Sharding/ShardedDatabaseContext.Subscriptions.cs:25]: it is the original name :-)

**Review by ppekrol — APPROVED** [review]
@karmeli87 can you check failing tests?

**Pull request merged**